# By design, to the furthest extent possible, Guido van Rossum made Python 3 a functional programming language.
## This means that programs are written by applying existing functions and composing new ones.
### This is why print() and python_version() in earlier notebooks are functions --- as opposed to, say, commands.

# Writing functions in Python is easy.
## This is not to say that making them correct and efficient is easy --- these are different subjects entirely.

# In this notebook, we will learn how to write functions in Python, as well as how to save and load them from text files called "modules", which makes maintaining and re-using them easy.

## At its simplest, a Python function needs a name only.

In [1]:
def testFunc(): # def is the reserved word in Python to create a function. The function name follows.
    pass

# after the function name, we immediately write the open parenthesis of the parentheses that hold the argument list,
# if any. after we write the close parenthesis, we end the line with a full colon.

# python indents the block of code associated with the function.
# in our example, we write a single statement, "pass".
# the pass statement does absolutely nothing, and is useful when troubleshooting code.

### Note that the Python interpreter writes an entry for the function into its <b>symbol table</b>.

In [3]:
%whos

Variable   Type        Data/Info
--------------------------------
testFunc   function    <function testFunc at 0x7f7ec06f14c0>


### testFunc() above takes no arguments and returns no values.
### We invoke it on the line below, as proof.

In [4]:
testFunc()

## Now we write a simple function that takes a single argument and returns a single value.

In [5]:
def get_cube_root(n):
    return n ** (1/3)

# double asterisk ** is the exponentiation operator.

## Python has 3 built-in numeric types: integer (int), floating-point (float), and complex (complex).

In [6]:
# all three built-in numeric types can be passed in as arguments.
print(get_cube_root(27))
print(get_cube_root(3.14))
print(get_cube_root(1+9j))

3.0
1.4643443505031195
(1.8422978087224575+0.9748950044722985j)


## Now let's take the code we wrote to find the gc-content percentage of a nucleotide string, and convert it into a function.

In [7]:
def get_gc_content_per(nuclStr):
    # the variable name we give to the argument is used in the function's block of code
    return 100*((nuclStr.count('g') + nuclStr.count('c'))/len(nuclStr))

# yes, it's as simple as this.

In [8]:
# and now we test
get_gc_content_per('gattacagattaca')

28.57142857142857

In [9]:
# the advantage of rendering the code as a function i hope is obvious.
# we can re-use it as often as we want on expected input ---
get_gc_content_per('acgtacgtaaaaaaa')

26.666666666666668

# Our get_gc_content_per() function works on expected input, but it is poorly written.
## For starters, even a first-yr high-school biology student would notice that it accepts only lower-case nucleotide strings, and that such strings can be either uppercase or lowercase.
## And our function and variable names are descriptive (as opposed to being, say, g() for the function name and x for the variable name). When we give functions and files and variables meaningful names in the contexts in which they're used, we call the code "self-documenting code".
## But our code is otherwise undocumented.
## Then there's no code to ensure that the users of our functions enter expected input only --- by which i mean strings composed only of uppercase or lowercase nucleotides.
### We will deal with these issues throughout the course.

# For now, let's save our function to a module.

### A module is a .py file that holds functions we write and use.
### The name we'll use for our class module is myModule.py
### In JupyterHub, we will upload our module to our notebook directory.
### In this environment, we will write and edit our module on our local machines, save it, then upload it to our notebook directory in order to use it.

In [10]:
# these magics allow the notebook to reload contents of edited modules
# into session memory
%load_ext autoreload
%autoreload 2

In [11]:
# copy/paste get_gc_content_per(nuclStr) into myModule.py
# save/upload myModule to notebook directory
# import myModule as mM
import myModule as mM

# try to invoke function in myModule
mM.get_gc_content_per('gattaca')

28.57142857142857

In [10]:
# edit/modify local copy of myModule.py
# save/upload myModule to notebook directory
# overwrite existing remote copy on JupyterHub
# del mM
del mM
# import myModule as mM
import myModule as mM

In [10]:
# try to invoke new function added to myModule


Variable             Type        Data/Info
------------------------------------------
get_cube_root        function    <function get_cube_root at 0x7fadc4732ca0>
get_gc_content_per   function    <function get_gc_content_per at 0x7fadc4732f70>
mM                   module      <module 'myModule' from '<...>home/jderry/myModule.py'>
testFunc             function    <function testFunc at 0x7fadc477fdc0>


# Improving our functions

## Adding type hints & docstring
### As of Python 3.5, the interpreter allows programmers to add type hints to their function arguments and return values.
<br>For mM.get_gc_content_per(), we modify the function head to read: <b>get_gc_content_per(nuclStr: str) -> float</b>

In [13]:
# we upload/overwrite our changes
# del mM
# import myModule as mM
# test
mM.get_gc_content_per?

### In Python, we document a function in a docstring (enclosed by 3 single or double quotes).
### A function docstring is added under the function head and includes: a type contract (if type hints aren't possible), a single-sentence explanation of what the function takes as input and returns as output; and at least one example of the function.
<br>For mM.get_gc_content_per(), we might add this docstring:</br>
<b><br>''' (str)->(float) # example of type contract.</br>
<br>Given a string of lowercase nucleotides, return the percentage of GC content.</br>
<br>>>>get_gc_content_per('gattaca')</br>
<br>   28.57142857142857</br></b>
<br>'''</br></b>

In [11]:
# we upload/overwrite our changes
# del mM
# import myModule as mM
# test
mM.get_gc_content_per?